File: framework/languages/dotnet/e2e/RuntimeMonitoring/Client/Support/ClientOptions.cs
namespace RuntimeMonitoring.Client.Support;

internal sealed record ClientOptions(
    string RedisEndpoint,
    string RedisKeyPrefix,
    string ServiceUrl,
    string ServiceChannelEndpoint,
    string ServiceBUrl,
    int ServiceBProcessId,
    string ServiceBChannelEndpoint,
    string ServiceBSpotRouterEndpoint,
    string ServiceBSpotPubEndpoint,
    string FilteredServiceUrl,
    string FilteredChannelEndpoint,
    string ThrowServiceUrl,
    string ThrowChannelEndpoint,
    string FilteredServiceProject,
    string ServiceProject,
    string ValidationHostProject,
    string Scenario,
    string LogDir)
{
    public static ClientOptions Parse(string[] args)
    {
        var values = new Dictionary<string, string>(StringComparer.Ordinal);
        for (var i = 0; i < args.Length; i++)
        {
            var key = args[i];
            if (!key.StartsWith("--", StringComparison.Ordinal))
                throw new ArgumentException($"Unexpected argument '{key}'.");

            if (i + 1 >= args.Length) throw new ArgumentException($"Missing value for '{key}'.");

            values[key] = args[++i];
        }

        string Get(string name)
        {
            return values.TryGetValue(name, out var value) && !string.IsNullOrWhiteSpace(value)
                ? value
                : throw new ArgumentException($"{name} is required.");
        }

        return new ClientOptions(
            Get("--redis-endpoint"),
            Get("--redis-key-prefix"),
            Get("--service-url"),
            Get("--service-channel-endpoint"),
            Get("--service-b-url"),
            int.Parse(Get("--service-b-process-id"), System.Globalization.CultureInfo.InvariantCulture),
            Get("--service-b-channel-endpoint"),
            Get("--service-b-spot-router-endpoint"),
            Get("--service-b-spot-pub-endpoint"),
            Get("--filtered-service-url"),
            Get("--filtered-channel-endpoint"),
            Get("--throw-service-url"),
            Get("--throw-channel-endpoint"),
            Get("--filtered-service-project"),
            Get("--service-project"),
            Get("--validation-host-project"),
            values.TryGetValue("--scenario", out var scenario) && !string.IsNullOrWhiteSpace(scenario)
                ? scenario
                : "all",
            Get("--log-dir"));
    }
}

File: framework/languages/node/e2e/RuntimeMonitoring/Client/Support/client-options.ts
export interface ClientOptions {
  readonly triggerUrl: string;
  readonly serviceUrl: string;
  readonly serviceBUrl: string;
  readonly throwServiceUrl: string;
  readonly redisEndpoint: string;
  readonly redisKeyPrefix: string;
  readonly serviceBChannelEndpoint: string;
  readonly serviceBSpotRouterEndpoint: string;
  readonly serviceBSpotPubEndpoint: string;
  readonly serviceMain: string;
  readonly serviceBConfig: string;
  readonly logDir: string;
  readonly scenario: string;
}

export function parseClientOptions(args: readonly string[]): ClientOptions {
  const values = new Map<string, string>();
  for (let i = 0; i < args.length; i += 1) {
    const key = args[i];
    if (!key.startsWith('--')) {
      continue;
    }
    if (i + 1 >= args.length) {
      throw new Error(`Missing value for ${key}.`);
    }
    values.set(key.slice(2), args[++i]);
  }
  return {
    triggerUrl: required(values, 'trigger-url'),
    serviceUrl: required(values, 'service-url'),
    serviceBUrl: required(values, 'service-b-url'),
    throwServiceUrl: required(values, 'throw-service-url'),
    redisEndpoint: required(values, 'redis-endpoint'),
    redisKeyPrefix: required(values, 'redis-key-prefix'),
    serviceBChannelEndpoint: required(values, 'service-b-channel-endpoint'),
    serviceBSpotRouterEndpoint: required(values, 'service-b-spot-router-endpoint'),
    serviceBSpotPubEndpoint: required(values, 'service-b-spot-pub-endpoint'),
    serviceMain: required(values, 'service-main'),
    serviceBConfig: required(values, 'service-b-config'),
    logDir: required(values, 'log-dir'),
    scenario: values.get('scenario') ?? 'all'
  };
}

function required(values: ReadonlyMap<string, string>, key: string): string {
  const value = values.get(key);
  if (value === undefined || value.length === 0) {
    throw new Error(`--${key} is required.`);
  }
  return value;
}
